File: frontend/src/types/reservation.ts
/**
 * Domain type contracts for reservations (a.k.a. bookings).
 *
 * A Reservation is the link between a User and a ParkingZone and represents
 * a locked-in spot. The backend enforces concurrency safety — if the zone is
 * full at the moment of creation the API returns HTTP 409, which the
 * reservationService surfaces as a typed error so the UI can render a
 * friendly toast.
 */

import type { ParkingZone } from "./zone";
import type { User } from "./auth";

export type ReservationStatus = "active" | "completed" | "cancelled";

/** Full reservation record as returned by GET /reservations/me and GET /reservations. */
export interface Reservation {
  id: number;
  user_id: number;
  zone_id: number;
  license_plate: string;
  status: ReservationStatus;
  zone?: ParkingZone;
  user?: User;
  created_at: string;
}

/** Payload used by POST /reservations to lock in a spot. */
export interface CreateReservationPayload {
  zone_id: number;
  license_plate: string;
}

File: backend/internal/models/reservation.go
package models

import "time"

// Reservation represents a parking spot reservation by a user.
// Valid Status values are: "active", "completed", "cancelled".
type Reservation struct {
	ID           uint        `gorm:"primaryKey;autoIncrement" json:"id"`
	UserID       uint        `gorm:"not null;index" json:"user_id"`
	User         User        `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE" json:"user,omitempty"`
	ZoneID       uint        `gorm:"not null;index" json:"zone_id"`
	Zone         ParkingZone `gorm:"foreignKey:ZoneID;constraint:OnDelete:CASCADE" json:"zone,omitempty"`
	LicensePlate string      `gorm:"type:varchar(15);not null" json:"license_plate"`
	Status       string      `gorm:"type:varchar(20);not null;default:'active'" json:"status"`
	CreatedAt    time.Time   `json:"created_at"`
	UpdatedAt    time.Time   `json:"updated_at"`
}
